### DISCUSSION TITLE
comparion between python and julia version of symbolic regression

### DISCUSSION THREAD

**Discussion by wenpw** [MilesCranmer/PySR#360] [Q&A]
Dear Miles,

Thank you very much for the powerful code.
Please help check the following problems.

Recently, I have compared the Pysr and the Julia version to fit the same formula:
y = 4*log(1+exp(2*pi*(x-100)/4)/2/x.

The codes are in the attached zip file.

Two serious problems were found:

1)
In Julia, it takes only few minutes to obtain the correct formula using only two cpus.
In Python, I have tried for more than one day using even twenty cpu, however, I still could not get the right result.
I have also tried different parameters and could still not find the correct answer.

May I ask why this happens, and how to improve the python script to get the correct results?

2)
In julia, hundreds of G RAM memory are needed in my servers for the running for example using numprocs=20.
In python, no large RAM memroy is needed in my test.
Is there any  way to release the memory during the running of Julia code?

Thank you very much in advance.

[log1p-v2.zip](https://github.com/MilesCranmer/PySR/files/11912030/log1p-v2.zip)

**Comment by MilesCranmer** [comment]
Dear @wenpw,
Thanks for posting. It looks like your search settings are different in Julia vs Python. Could you please try with the same settings?

Also can  you paste the code into your answer directly rather than through a zip file? 
Thanks,
Miles

**Reply by wenpw** [reply]
Dear @MilesCranmer:

Thank you for the quick reply.
In fact, I have tried many different settings for these two version.
And the problems are smilar.

Here are the codes:

1) The python version:
```python

import numpy as np
import sympy

def wong(expe,vb,rb,curv):
    fwong =  curv  * np.log(1.0 +np.exp(2.*np.pi*(expe-vb)/curv)) /2./expe
    return fwong

x0 = 85+np.random.rand(300,1)*150.
x1 = 100.
x2 = 10.
x3 = 4.

curv = x3*np.ones((300,1))
rb = x2*np.ones((300,1))
vb = x1*np.ones((300,1))
expe = x0
y = wong(expe,vb,rb,curv)

X = expe

from pysr import PySRRegressor

model = PySRRegressor(
    procs=400,
    populations=200,
    # ^ 2 populations per core, so one is always running.
    population_size=500,
    # ncyclesperiteration=550, 
    # ^ Generations between migrations.
    niterations=10000000,  # Run forever
    early_stop_condition=(
        "stop_if(loss, complexity) = loss < 1e-6 && complexity < 10"
        # Stop early if we find a good and simple equation
    ),
    timeout_in_seconds=60 * 60 * 24 * 20,
    maxsize=20,
    binary_operators=["*", "+", "-", "/"],
    unary_operators=[
        "exp",
        "log1p",
    ],
    loss="loss(x, y) = abs(x - y)^2 + 30000*(log(abs(x)) - log(abs(y)))^2",
    # loss="loss(x, y) =  (log(abs(x)+1e-48) - log(abs(y)+1e-60))^2",
    # ^ Custom loss function (julia syntax)
    progress=True,
    warm_start=True,
    turbo=False,
)

model.fit(X, y)

print(model)

print(model.predict(X))

```

2) the Julia version:

```julia
using SymbolicRegression, SymbolicUtils

expe = 85.0f0 .+ rand(Float64, 1, 300) .*150#.0f0
curv = expe[1,:] .- expe[1,:] .+ 4.0f0
rb = expe[1,:] .- expe[1,:] .+ 10.0f0
vb = expe[1,:] .- expe[1,:] .+ 100.0f0


X=expe
y =  curv[1,:] .*  log.(1.0f0 .+ exp.(2.0f0 * pi .*(expe[1,:]  .-vb[1,:] )./ curv[1,:])  ) ./2.0f0 ./expe[1,:]


early_stop(loss, c) = ((loss <= 1e-10) && (c <= 10))
my_loss(x, y) = abs(x - y)^2 + 30000*(log(abs(x)) - log(abs(y)))^2

options = SymbolicRegression.Options(;
    binary_operators=[+, *, /, -], 
    unary_operators=[exp, log1p], 
    npopulations=200,
    early_stop_condition=early_stop,
    enable_autodiff=true,
    elementwise_loss=my_loss,
)

hall_of_fame = EquationSearch(
    X, y;
    niterations=400000000, 
    options=options, 
    numprocs=2,
    parallelism=:multiprocessing,
)

dominating = calculate_pareto_frontier(hall_of_fame)

trees = [member.tree for member in dominating]

tree = trees[end]
output, did_succeed = eval_tree_array(tree, X, options)

eqn = node_to_symbolic(dominating[end].tree, options)
println("Complexity\tMSE\tEquation")

for member in dominating
    complexity = compute_complexity(member.tree, options)
    loss = member.loss
    string = string_tree(member.tree, options)
    println("$(complexity)\t$(loss)\t$(string)")
end

```

**Comment by MilesCranmer** [comment]
(e.g., you requested 400 cores in PySR, but only 2 cores in SymbolicRegression.jl)

**Reply by wenpw** [reply]
Year, the first serious problem is:

Even when I use 400 cores in Pysr, I could not get the correct answer for more than 1 day's running.
Using 2 cores, the answer could still not be obtained.

In julia, 2 cores are encough within few minuits but with huge RAM memories.

**Comment by MilesCranmer** [comment]
> Even when I use 400 cores in Pysr, I could not get the correct answer for more than 1 day's running.

Just to be clear – if you only have 20 cores, then you should not set a higher number than 20 for procs. Even if it is working, that would slow things down quite a bit as it overload the CPU scheduler.

**Comment by MilesCranmer** [comment]
I noticed you have `parallelism=:multiprocessing` in the Julia version. Could you try using that for Python as well? In the Python version, that would be `multithreading=False`.

**Reply by wenpw** [reply]
I have tried this setting and run for about one hour, and nothing improved yet.
I had also tried adjusting many other parameters before, and could not get the correct answer.

You may try to run this short Python version of the code and test it futher.

**Reply by MilesCranmer** [reply]
What do you see as the output with multithreading=False?

Again I should really emphasize that you need to run with the same settings in PySR and SymbolicRegression.jl. Otherwise it is impossible to know what is the cause of the slowdown.

**Reply by wenpw** [reply]
As mentioned above, with multithreading=False included the results does not change much except that the occupied memories are large now.
The correct answer could still not be found.

I would also like to to run with the same settings in PySR and SymbolicRegression.jl. 
But I think some hidden parameters inside the source code of the programs which are not listed in above scripts are not the same yet for these two versions.

**Reply by MilesCranmer** [reply]
Can you paste the code exactly as you are running it? And also give me details on your system, the PySR version, the Julia version, the SymbolicRegression.jl version. For example, there is a chance you are running a different SymbolicRegression version than the one PySR is using.

For collecting a comparison, please use a minimal example. Do not change the default parameters in either PySR or Julia. For the Julia version run with `-O3 --threads=auto` to mimic the PySR launch.

> But I think some hidden parameters inside the source code of the programs which are not listed in above scripts are not the same yet for these two versions.

This should not be the case. Some parameters have different names (will fix eventually) but the actual parameter values should be 1:1. If they are not 1:1 that is a bug and I will fix asap.

**Reply by wenpw** [reply]
Here are the exactly code I am running:

1 Pysr:

```python
import numpy as np
import sympy

def wong(expe,vb,rb,curv):
    fwong =  curv  * np.log(1.0 +np.exp(2.*np.pi*(expe-vb)/curv)) /2./expe
    return fwong

x0 = 85+np.random.rand(300,1)*150.
x1 = 100.
x2 = 10.
x3 = 4.

curv = x3*np.ones((300,1))
rb = x2*np.ones((300,1))
vb = x1*np.ones((300,1))
expe = x0
y = wong(expe,vb,rb,curv)

X = expe

from pysr import PySRRegressor

model = PySRRegressor(
    procs=1,
    niterations=10000000,  # Run forever
    early_stop_condition=(
        "stop_if(loss, complexity) = loss < 1e-6 && complexity < 10"
    ),
    timeout_in_seconds=60 * 60 * 24 * 20,
    maxsize=20,
    binary_operators=["*", "+", "-", "/"],
    unary_operators=[
        "exp",
        "log1p",
    ],
    loss="loss(x, y) = abs(x - y)^2 + 30000*(log(abs(x)) - log(abs(y)))^2",
    progress=True,
    multithreading=False
)

model.fit(X, y)

print(model)

print(model.predict(X))
```

2: Julia version

```julia
using SymbolicRegression, SymbolicUtils

expe = 85.0f0 .+ rand(Float64, 1, 300) .*150#.0f0
curv = expe[1,:] .- expe[1,:] .+ 4.0f0
rb = expe[1,:] .- expe[1,:] .+ 10.0f0
vb = expe[1,:] .- expe[1,:] .+ 100.0f0


X=expe
y =  curv[1,:] .*  log.(1.0f0 .+ exp.(2.0f0 * pi .*(expe[1,:]  .-vb[1,:] )./ curv[1,:])  ) ./2.0f0 ./expe[1,:]


early_stop(loss, c) = ((loss <= 1e-10) && (c <= 10))
my_loss(x, y) = abs(x - y)^2 + 30000*(log(abs(x)) - log(abs(y)))^2

options = SymbolicRegression.Options(;
    binary_operators=[+, *, /, -], 
    unary_operators=[exp, log1p], 
    #npopulations=200,
    early_stop_condition=early_stop,
    #enable_autodiff=true,
    elementwise_loss=my_loss,
)

hall_of_fame = EquationSearch(
    X, y;
    niterations=400000000, 
    options=options, 
    #numprocs=2,
    #parallelism=:multiprocessing,
)

dominating = calculate_pareto_frontier(hall_of_fame)

trees = [member.tree for member in dominating]

tree = trees[end]
output, did_succeed = eval_tree_array(tree, X, options)

eqn = node_to_symbolic(dominating[end].tree, options)
println("Complexity\tMSE\tEquation")

for member in dominating
    complexity = compute_complexity(member.tree, options)
    loss = member.loss
    string = string_tree(member.tree, options)

    println("$(complexity)\t$(loss)\t$(string)")
end
```

----------------------------------------------------------

> For collecting a comparison, please use a minimal example. Do not change the default parameters in either PySR or Julia. For the Julia version run with -O3 --threads=auto to mimic the PySR launch.

I have tried my best to use default parameters in the above code. And I have run with -O3 --thread=auto in the running.

> And also give me details on your system, the PySR version, the Julia version, the SymbolicRegression.jl version. For example, there is a chance you are running a different SymbolicRegression version than the one PySR is using.

My Python version: 3.8.8; Julia version: 1.9.0-rc3; Pysr version: 0.14.2; SymbolicRegression.jl version: v0.19.0.

> This should not be the case. Some parameters have different names (will fix eventually) but the actual parameter values should be 1:1. If they are not 1:1 that is a bug and I will fix asap.

Thank you for above comments. Please help check the above codes.

**Reply by MilesCranmer** [reply]
Thanks, this is very helpful. I'm looking now. The one note I will make is that your Julia version you are using Float64, whereas the PySR version you are using Float32. (SymbolicRegression.jl figures out the type based on the array you pass it; whereas PySR this is set explicitly by `precision`).

Also what is your operating system?

**Reply by MilesCranmer** [reply]
Why are you using
```python
 procs=1,
```
?

This will limit the PySR-launched search to only one process.

**Reply by wenpw** [reply]
> The one note I will make is that your Julia version you are using Float64, whereas the PySR version you are using Float32. (SymbolicRegression.jl figures out the type based on the array you pass it; whereas PySR this is set explicitly by precision).

I tried following calculations by using Float64 in the Pysr version, and the results do not improve.

```python
import numpy as np
import sympy

def wong(expe,vb,rb,curv):
    fwong =  np.float64(curv  * np.log(1.0 +np.exp(2.*np.pi*(expe-vb)/curv)) /2./expe)
    return fwong

x0 =  np.float64(85+np.random.rand(300,1)*150.)
x1 =  np.float64(100.)
x2 =  np.float64(10.)
x3 =  np.float64(4.)

curv = np.float64(x3*np.ones((300,1)))
rb = np.float64(x2*np.ones((300,1)))
vb = np.float64(x1*np.ones((300,1)))
expe = np.float64(x0)
y = wong(expe,vb,rb,curv)

X = expe

from pysr import PySRRegressor

model = PySRRegressor(
    procs=10,
    niterations=10000000,  # Run forever
    early_stop_condition=(
        "stop_if(loss, complexity) = loss < 1e-6 && complexity < 10"
    ),
    timeout_in_seconds=60 * 60 * 24 * 20,
    maxsize=20,
    binary_operators=["*", "+", "-", "/"],
    unary_operators=[
        "exp",
        "log1p",
    ],
    loss="loss(x, y) = abs(x - y)^2 + 30000*(log(abs(x)) - log(abs(y)))^2",
    progress=True,
    multithreading=False
)

model.fit(X, y)

print(model)

print(model.predict(X))
```


>  procs=1,

This is an example. I have also tried procs =10 or larger value. The results do not change much for finding the correct formula.

> Also what is your operating system?

I have ran above example code in Ubuntu 22.04.2 LTS and windows 10.

**Reply by MilesCranmer** [reply]
Thanks for your continued reporting. Can you share details on your measurements of execution speed? For example, a screenshot of `htop`? The Ubuntu data will be more useful for me.

For the record I'm not able to reproduce your measurements; your PySR example runs fast for me on both macOS and linux:

<img width="1667" alt="Screenshot 2023-07-02 at 9 55 27 PM" src="https://github.com/MilesCranmer/PySR/assets/7593028/80890129-4923-415f-b43b-546ae4677e9f">

**Reply by MilesCranmer** [reply]
Also note that the way you use 64-bit precision in PySR is by passing `precision=64`, **not** by passing data of 64-bit type. This is one difference between PySR and SymbolicRegression.jl's frontends.

**Reply by wenpw** [reply]
Thank you for your patience.

> Can you share details on your measurements of execution speed? For example, a screenshot of htop? The Ubuntu data will be more useful for me.

Here is the htop data in my Ubuntu server:
![微信图片_20230703103445](https://github.com/MilesCranmer/PySR/assets/92563457/e3c61150-6ff1-4a55-a7fe-d1ac98e31566)

> For the record I'm not able to reproduce your measurements; your PySR example runs fast for me on both macOS and linux:

Do you mean that you could obtain the corrrect answer by Pysr on macOS and Linux very quickly?

**Reply by wenpw** [reply]
> Also note that the way you use 64-bit precision in PySR is by passing precision=64, not by passing data of 64-bit type. This is one difference between PySR and SymbolicRegression.jl's frontends.

Thank you for this reminding.

**Reply by MilesCranmer** [reply]
Wait I am confused. I thought you mentioned you were not able to get multithreading working? It looks like it is using multiple cores there, no?

**Reply by wenpw** [reply]
Yeah, there is no problems using multiple cores. The Problem is using Pysr, no matter how I adjust the parameters, I could not get the correct answer. Using julia, correct answer could be obtained very soon even with only 1 cpu.
